### ISSUE TITLE
Adding considerations for which camera platform could be right for which use case

### ISSUE THREAD

**Issue by scstraus** [home-assistant/home-assistant.github.io#17781]
### Feedback

Based on encouragement from @uvjustin [here](https://community.home-assistant.io/t/i-tried-all-the-camera-platforms-so-you-dont-have-to/222999/175?u=scstraus), I'm going to try to summarize some of the things from my [I tried all the camera platforms so you don't have to](https://community.home-assistant.io/t/i-tried-all-the-camera-platforms-so-you-dont-have-to) thread to provide a structure for what some good things would be to have in the camera documentation to help guide people to the correct decisions. In that thread I've made some observations and guesses as to causes without really understanding the underlying workings well enough to make concrete statements that will likely apply to all environments. Hopefully you can fill in some of the below in an informed enough way that I can pull it together into an actual PR for the docs.

Things that would help to guide people to good deicisons regarding cameras starting from the main camera page:

1. A list of all the camera platforms with links to each of them
2. A short description of the advantages, philosophy, and requirements for each camera platform such as:
* It's relative server side CPU usage (high/medium/low)
* It's relative client side CPU usage (high/medium/low)
* Bandwidth requirements from hass to clients (high/medium/low in summary but a calculator would be nice in the detailed docs)
* Ability to handle things like dropped frames and sporatic connections (high/medium/low in summary, but more detail in each platform)
* Client requirements to display video properly (more modern browser, etc.)
* Supported codecs and protocols on camera side
* General expectations around lag and loading times and anything which can influence that (eg. frequency of keyframes in source RTSP stream when using HLS)
* What influence turning the stream component on or off will have on the specific type of camera stream
* Maybe some general considerations to consider about IP cameras (some of them might allow only one stream at a time of a specific type so you might have to use a proxy such as the hass camera proxy or live555 proxy, etc).

Some of this could be a comparison table on the main camera page and some of it could be included in more detail on the camera page itself. While I have included my observations for most of these things in the [afformetioned thread](https://community.home-assistant.io/t/i-tried-all-the-camera-platforms-so-you-dont-have-to/222999/175?u=scstraus), I have no idea of the underlying causes of my observations and how many of them simply relate to my specific configuration.

Some general troubleshooting guidelines would be good too, here's what I had in my thread, probably too wordy for the actual docs, but I could rewrite or add in what you think best practices are.

> 
> 1) Overloaded CPU (/GPU?) on your server: Any time you are taking an h.264 stream and displaying it as MJPEG, there is one ffmpeg instance being created for each camera being displayed on each frontend device. 4 cameras being displayed on 3 devices? 12 instances. Each of these instances uses something like 8-10% of my CPU so displaying more than 8 of these cameras is impossible on my old core duo mac mini. I’ve never tried to see what was possible with GPU, my understanding from an FR is that the NUC hass.io 2 image has the ability to use the intel integrated GPU, so people running intel NUC’s with hass.io 2 may be utilizing GPU resources for this, so it would be good to check those too.
> 
> Solutions:
> 
>     Activate streaming component so that your cameras display as HLS instead of MJPEG (though you will pay for this with lag, which in my case is about 10 seconds.
>     Use proxy camera component to re-serve the stream. This magically seems to not use much of any CPU, though it does also make the camera stream less reliable so that sometimes it still won’t come up.
>     Buy a better server :wink:
> 
> 2) Limited number of streams available from your camera
> 
> Some cameras will only allow one stream of a certain type or one stream at a time from a type of device. Keep in mind that generally homeassistant is not re-casting your streams. When you open a stream to display on the frontend, it opens that stream from the camera or DVR that you’ve configured. It means you are likely to have many streams going to the camera and/or DVR at once if you have multiple displays (like I do with my tablets).
> 
> A couple examples of this that I’ve encountered are that
> 
>     My Hikvision cameras are only able to serve one MJPEG stream at a time, so the second device to try opening will get a blank card. They seem happy to stream many h.264 camera streams, though.
> 
>     My reolink camera will only allow streaming via the app or via the stream that the reolink beta component uses at a time. So, as soon as I open the reolink app, all the homeassistant streams vanish.
> 
> Solutions:
> 
>     Try using h.264 instead of MJPEG
>     Account for everything that’s using the camera and try to close clients to see if it allows more clients to connect.
> 
> 3) Overloaded CPU on your cameras
> 
> This is one I realized too late to really assess it properly, as I didn’t have a sensor for it until I got the ONVIF cameras going. But basically cameras have very limited CPU resources, and if you are running some of the features like motion detection, line crossing, tampering, camera rotation in addition to certain types of apparently demanding streams like MJPEG or to a lesser extent the proxy camera component, you can overload the CPU. That being said, mine still seem to work pretty well when almost maxed out but I’m sure it caused some of my problems.
> 
> Solutions:
> 
>     Avoid MJPEG cameras. For whatever reason they have high CPU load
>     Avoid rotating your cameras. This causes me some additional load
>     Avoid proxy camera. Adds some small load
>     Don’t use camera motion detection features (you can do this on the synology or DVR for example)
> 
> 4) Overloaded browser on the frontend
> 
> I see this sometime on my kindle fire tablets if I am trying to display 4x 1080p stream cameras with live view enabled. But instead of not displaying the camera, fully kiosk browser just crashes, so a bit different. But your frontend devices do have limits. One other browser limit that’s common across most desktop browsers (as far as I understant) is the limitation of not being able to open more than 6 connections to a domain. So if you are displaying 5 camera feeds, you will not be able to open a second browser window to your hass instance without having camera problems.
> 
> Solutions:
> 
>     Use a less demanding camera type (non live stream, mjpeg appears easier to stream on the frontend than ALS using stream).
>     Get a stronger frontend device. A modern iOS device seems to be able to keep up with just about anything I’ve tried, compared to a low powered Kindle Fire device. YMMV.
>     Don’t open multiple hass browser windows on the same computer
> 

Anyhow, if you can give me some confirmations as people who understand how these camera platforms work to some of my observations, I will compile them and open a PR with the text. That's one thing I'm pretty good at. Feel free to tag anyone you feel can help.

### URL

https://www.home-assistant.io/integrations/camera/

### Version

2021.5.1

**Comment by uvjustin** [comment]
What specific things do you need confirmed? Bandwidth estimates can be found from a quick google search.
I think keeping it simple is good. The high/medium/low thing makes sense. Maybe something like:
```

>                               stream      ffmpeg     mjpeg
> camera processing             low         low        high
> HA server processing          low         high       low
> client processing             medium      low        low
> camera<->HA server bandwidth  low         low        high
> HA server<->client bandwidth  low         high       high
> latency                       high        low        low

```

**Comment by scstraus** [comment]
These are the things I wanted confirmed:


> 1. A list of all the camera platforms with links to each of them
> 2. A short description of the advantages, philosophy, and requirements for each camera platform such as:
> * It's relative server side CPU usage (high/medium/low)
> * It's relative client side CPU usage (high/medium/low)
> * Bandwidth requirements from hass to clients (high/medium/low in summary but a calculator would be nice in the detailed docs)
> * Ability to handle things like dropped frames and sporatic connections (high/medium/low in summary, but more detail in each platform)
> * Client requirements to display video properly (more modern browser, etc.)
> * Supported codecs and protocols on camera side
> * General expectations around lag and loading times and anything which can influence that (eg. frequency of keyframes in source RTSP stream when using HLS)
> * What influence turning the stream component on or off will have on the specific type of camera stream
> * Maybe some general considerations to consider about IP cameras (some of them might allow only one stream at a time of a specific type so you might have to use a proxy such as the hass camera proxy or live555 proxy, etc).

For bandwidth, I agree you can find it with a google search, but to actually do a calculation, you'd need to know what's getting converted to what at what settings. That is not information that's currently available for each camera platform. So maybe that's a good place to start.

Your table is a good start. One question I would have is what it means to have ffmpeg with stream on vs ffmpeg with stream off, and mjpeg with stream on and off, etc. Stream can be combined with any camera platform as it's a system-wide setting and I have noticed that each camera platform reacts somewhat differently to having stream turned on and off. I have no idea what's happening in the backend in those cases. I opened an FR a long time ago asking to be able to enable/disable stream on a camera by camera or at least camera platform by camera platform basis but never made it too far with that.

**Comment by uvjustin** [comment]
I'm only familiar with the stream component and the generic component so I can't speak definitively on the others without digging in to the code. Maybe someone else can help out with those.
I'm not sure a complete taxonomy of all camera platforms makes sense, since several of the camera platforms are just hardware specific ones. However, I do think it would make sense to clarify the distinction between the "meta" platforms, since there are several that can be used seemingly interchangeably but with different results. I think you have already identified the relevant ones in your forum topic (I would exclude synology since it is hardware specific).
For bandwidth, most of the variability doesn't come from any settings we use but from the camera itself. Besides codec and resolution, the bandwidth might vary a bit depending on frame rate and i-frame frequency. I don't have any more insight on the actual bandwidth besides what I mentioned earlier in broad strokes. I think you can use these links as reference (from the first page of a google search):
https://reolink.com/ip-camera-bandwidth-calculation/
http://www.stardot.com/bandwidth-and-storage-calculator
I think the answers to all the topics you mentioned are actually found in your forum thread except:

> What influence turning the stream component on or off will have on the specific type of camera stream

I'm actually not sure how some of the other "meta" components use stream either, Let's see if someone else can answer that - otherwise I'll try and take a look at those components when I have a bit more time.

**Comment by frenck** [comment]
> A list of all the camera platforms with links to each of them

✅  <https://www.home-assistant.io/integrations/#camera>

**Comment by frenck** [comment]
> A short description of the advantages, philosophy, and requirements for each camera platform such as

IMHO this does not belong on our website or on an overview page within Home Assistant. Overview pages are nice, but never cover all use cases and content; and more often are hard to read/impossible to view on mobile. Besides, we provide integration that match the device used and should therefore provide the best experience with your device.

If that isn't the case, a report should be filed.

A lot of the things listed in your list are camera specific, and not a concern to Home Assistant at all.

**Comment by scstraus** [comment]
To be honest, that was kind of the response I was expecting which is why I didn't attempt the documentation route in the first place and just went with a forum post. 

I think in general you are right, if you've got a camera platform that matches a brand and works perfectly, that would be ideal.

Real life isn't so simple unfortunately. In reality with my hikvision cameras, I have about 6-8 options, each with their strengths and weaknesses. Despite hikvision being perhaps the biggest IP camera manufacturer in the world, they do not have a branded integration listed on your integrations page. And even if they did, there is no way that it could be the best possible integration for all the use cases, because to do that, it would have to encompass all the other integrations such as FFmpeg, proxy, onvif, stream, mjpeg, WebRTC, etc, and despite my feeling that that's what needs to happen to hass camera support, that's not where we are.  

Which creates a very complex and confusing landscape for anyone trying to do cameras in hass. Most attempts to use camera support in hass end poorly.  I have spent literally hundreds if not thousands of hours trying to get mine working the way I want and still have largely not achieved that. 

The right way would be to combine all the options into a super camera integration and then make a branded camera integration for each camera type which utilizes all the possibile options and enables them as checkboxes in the UI, but in the mean time we could at least stanch the bleeding with a bit better documentation. Someone looking at that list of camera integrations will be very unlikely to choose the right one to fit their requirements the first time.

In case you need to see more of the complexity for real world use cases I'm talking about, read my thread ["I tried all the camera platforms so you don't have to"](https://community.home-assistant.io/t/i-tried-all-the-camera-platforms-so-you-dont-have-to). If you've looked at it and the dev team agrees we don't need it in the docs, I will not waste time on the excercise.

**Comment by frenck** [comment]
I see what you want to achieve, but let's say we now start talking about weather platforms, or motion sensors... heck light bulbs...

Why camera's and not those others?

It is great you did research and thanks for sharing that, but I don't see how that would fit in the docs for now and the future in a maintainable and more universal way.

**Comment by scstraus** [comment]
My short answer to that question would be "you don't need additional information to make a weather, motion, or light bulb platform work." With those platforms you can largely just choose the one that matches your brand and standards. With cameras you often can't make it work without this information due to the fact there are too many choices that work for your specific device, and it's known only to those who have tried and failed and the developers.

But seeing as we have bumped up against philosophy here, I will close this and abandon the project unless @uvjustin has any other ideas.

**Comment by frenck** [comment]
I do not agree with that statement. I can hook up my ikea blubs to Home Assistant via:

The tradfri integration
The Hue integration
Via the deconz integration
Or use ZHA as the integration
Or use zigbee2mqtt with the mqtt integration.

Blubs are no different. All come with different pros and cons and make your bulb behave differently. For example, hue is polling.


Wat actually would work better on a longer term, is having an actual device database. Allowing to advise per device on what best matches the actual device.

**Comment by scstraus** [comment]
Lighting isn't a big focus of mine so I may be out of my depth here, but I think the difference is that all those platforms will allow you to turn on and off your lightbulbs, while many of the camera platforms will not display an adequate picture. The picture will freeze, drop out for minutes or hours at a time, smear, have unusable framerate or lag, crash your browser, bring your hass server's CPU to it's knees and grind the whole hass instance to a halt, take 20 seconds to initially appear, etc. 

I think with lights, you could easily choose any of those options and they would work for at least the basic light use case of  turning them on and off. Not the case for cameras. And the failure cases are far more spectacular. I doubt any of those light platforms can completely freeze the hass instance except in some extreme and rare edge case (which ffmpeg cameras are not).

**Comment by uvjustin** [comment]
> The right way would be to combine all the options into a super camera integration and then make a branded camera integration for each camera type which utilizes all the possibile options and enables them as checkboxes in the UI, but in the mean time we could at least stanch the bleeding with a bit better documentation. Someone looking at that list of camera integrations will be very unlikely to choose the right one to fit their requirements the first time.

I don't think there is a "right way" in HA - I think by its nature HA provides many ways to skin a cat. This is probably different to other (commercial) ecosystems like Smartthings, where there's one way to do everything but a lot less flexibility. As @frenck pointed out, the same multiple way approach applies to lighting, weather, etc. (BTW, if there was one "right way" I would advocate for just using stream and getting rid of the other components, but then you would lose the flexibility of other components that might work better for you.)

As @hunterjm previously mentioned in your forum topic, at the end of the day HA is a home automation platform and not a video platform. You can see this statement from @balloob himself at https://github.com/home-assistant/architecture/issues/113#issuecomment-443872128 . That's not to say we should not improve documentation or make improvements to some components where we can, but maybe your idea of what HA should offer video-wise is just outside the scope of HA.

That said, I still think there is room to improve the documentation with pointers in the right direction. I think stream (with generic, ONVIF, or a device specific integration) makes the most sense for users to start with as it is the most parsimonious with resources (many of the issues you mention such as smearing, CPU, or freezing problems are a result of using too many resources) while being compatible with casting. If you start using ffmpeg or mjpeg, things can probably get dicier (caveat - I don't know much about those integrations). Maybe the right approach is to just tell people to avoid ffmpeg or mjpeg unless they know what they are doing, while improving documentation on those components? Or maybe it's better to just make a new forum guide/howto topic which covers the stuff you've covered in your existing forum topic, but more definitively.

**Comment by scstraus** [comment]
>I don't think there is a "right way" in HA - I think by its nature HA provides many ways to skin a cat. 

By "right way" I meant to include every single possibile way of utilizing each camera directly into it's own branded camera integration, so you could use stream of FFMPEG or MJPEG without ever touching any camera integration other than (in my case Hikvision). You would just check the box for the option you wanted like "transcode my video on the server side" for FFMPEG, or "Use MJPEG instead of RTSP" for MJPEG, etc. In that way you'd get all the ways to "skin the cat" that would actually work on your hardware.. But it would be difficult to do because it would have to be matched to the capabilities of each individual camera.. 

But this philosophy could  also just be applied to one universal camera platform without trying to match to brand (matching to each brand I think is too big of a job and creates too much fragmentation and semi-orphaned platforms).

>HA is a home automation platform and not a video platform. You can see this statement from @balloob himself at home-assistant/architecture#113 (comment)

I've never really argued with that philosophy. I am not really looking for a video platform. Whatever we want to call it, hass already has all the features I need for cameras (and has had for years already). It's just a little hard to make them work the way I want them to.

>Maybe the right approach is to just tell people to avoid ffmpeg or mjpeg unless they know what they are doing

I think the matrix we started drafting and a couple warnings would help a hell of a lot. We don't have to boil the ocean. But I guess we could try to make some cookbook or something too.. I don't really know where to put something like that so people would see it, which is why I just made a forum post with my learnings.

**Comment by scstraus** [comment]
Looks like this won't go anywhere so I'm closing. Maybe someday I will wrap my head around this well enough and have the time to create some sort of flowchart for choosing which camera platform will work best for which use case. If that happens, I'll just put it up on the forums.
